util/device_sw_utils/extract_sw_logs.py

```diff
@@ -0,0 +1,123 @@
+#!/usr/bin/env python3
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+"""Script to convert logs placed in a given section into a parse-friendly database.
+
+The tool uses the readelf utility to extract const string messages placed in a
+given section of the ELF file generated for the device sw application. It appends
+the offset address of each string constant with the base address of the section
+to get the compete address of the log. The SW log monitor looks up this database
+to extract the logs at the right address and prints it with the appropriate severity.
+"""
+
+import argparse
+import logging as log
+import os
+import re
+import subprocess
+import sys
+
+
+def extract_sw_logs(readelf, section, input, output):
+    # Get the base address of the section
+    addr = 0
+    cmd = "readelf -S " + input + " | grep -F " + section
+    cmd_output = subprocess.run(cmd,
+                                check=True,
+                                shell=True,
+                                stdout=subprocess.PIPE,
+                                stderr=subprocess.STDOUT)
+    section_attrs = cmd_output.stdout.decode('utf-8')
+    # Output is of thie format:
+    # [10] .logs             PROGBITS        00000000 005000 000149 00 ...
+    # The start addr is the first attribute after 'PROGBITS'.
+    match = re.match(r"\s*\[.*\]\s+" + section + \
+                     "\s+\w+\s+([a-fA-f0-9]+)\s+.*", section_attrs)
+    if match:
+        addr = int(match.groups()[0].strip(), 16)
+    else:
+        log.info(
+            "Failed to extract the start addr of \"%s\" in \"%s\" file."
+            "Using addr = 0", section, input)
+
+    # Extract the logs from the section using the readelf utility.
+    cmd = [readelf, '-p', section, input]
+    cmd_output = subprocess.run(cmd,
+                                check=True,
+                                stdout=subprocess.PIPE,
+                                stderr=subprocess.STDOUT)
+    # Decode log texts into utf-8 string, replace bytes that do not conform.
+    text = cmd_output.stdout.decode('utf-8',
+                                    errors='replace').strip().split("\n")
+
+    # Each line of the log looks like this:
+    #   [   a4] Hello World!
+    #   OR
+    #   [   a4] I "hello_world.c" 25 0 "Hello World!"
+    #
+    # Convert this to:
+    # addr: 900000a4
+    # severity: I
+    # file: hello_world.c (or unknown_file)
+    # line: 25 (or 0)
+    # nargs: 0
+    # format: Hello World
+    result = ""
+    for line in text:
+        match = re.findall(r"^\s*\[\s*([a-fA-F0-9]+)\]  (.*)$", line)
```

Comment:
- Assign this to a regex object first to avoid re-compiling the regex for each line.
- You can `re.fullmatch()` here and avoid the begin/end markers; or use `re.search()`